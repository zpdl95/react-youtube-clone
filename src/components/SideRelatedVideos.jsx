import React from 'react';
import { useQuery } from '@tanstack/react-query';
import { useYoutubeApi } from '../context/YoutubeApiContext';
import VideoCard from './VideoCard';

export default function SideRelatedVideos({ id }) {
  const { youtube } = useYoutubeApi();
  const {
    isLoading,
    isError,
    data: videos,
  } = useQuery({
    queryKey: ['related', id],
    queryFn: ({ queryKey }) => {
      const [_, id] = queryKey;

      return youtube.relatedVideos(id);
    },
    staleTime: 1000 * 60 * 10,
  });

  return (
    <>
      {isLoading && <p>Loading...</p>}
      {isError && <p>Error!</p>}
      {videos && (
        <ul className='flex flex-col gap-2 mb-4'>
          {videos.map((video) => (
            <VideoCard key={video.id} video={video} type='list' />
          ))}
        </ul>
      )}
    </>
  );
}
