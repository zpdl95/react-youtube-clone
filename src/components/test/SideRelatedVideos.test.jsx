import React from 'react';
import {
  render,
  screen,
  waitFor,
  waitForElementToBeRemoved,
} from '@testing-library/react';
import { vi } from 'vitest';
import { Route } from 'react-router-dom';
import { fakeVideos } from '../../test/videos';
import { withAllContexts, withRouter } from '../../test/utils';
import SideRelatedVideos from '../SideRelatedVideos';

describe('SideRelatedVideos', () => {
  const fakeYoutube = { relatedVideos: vi.fn() };

  afterEach(() => fakeYoutube.relatedVideos.mockReset());

  it('SideRelatedVideos snapshot', async () => {
    fakeYoutube.relatedVideos.mockImplementation(() => fakeVideos);

    const { asFragment } = renderSideRelatedVideos();

    await waitForElementToBeRemoved(screen.queryByText('Loading...'));
    expect(asFragment()).toMatchSnapshot();

    // await waitFor(() => screen.getByRole('list'));
    // expect(asFragment()).toMatchSnapshot();
  });

  it('renders related videos correctly', async () => {
    fakeYoutube.relatedVideos.mockImplementation(() => fakeVideos);

    renderSideRelatedVideos();

    expect(fakeYoutube.relatedVideos).toHaveBeenCalledWith('id');
    await waitFor(() => {
      expect(screen.getAllByRole('listitem')).toHaveLength(fakeVideos.length);
    });
  });

  it('renders loading', () => {
    fakeYoutube.relatedVideos.mockImplementation(() => fakeVideos);

    renderSideRelatedVideos();

    expect(screen.getByText('Loading...')).toBeInTheDocument();
  });

  it('renders error', async () => {
    fakeYoutube.relatedVideos.mockImplementation(() => {
      throw new Error('error');
    });

    renderSideRelatedVideos();

    await screen.findByText('Error!');
  });

  function renderSideRelatedVideos() {
    return render(
      withAllContexts(
        withRouter(<Route path='/' element={<SideRelatedVideos id='id' />} />),
        fakeYoutube
      )
    );
  }
});
