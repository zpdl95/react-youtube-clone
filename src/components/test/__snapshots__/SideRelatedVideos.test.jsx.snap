// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SideRelatedVideos > SideRelatedVideos snapshot 1`] = `
<DocumentFragment>
  <ul
    class="flex flex-col gap-2 mb-4"
  >
    <li
      class="flex gap-2 "
    >
      <img
        alt="title"
        class="cursor-pointer w-40 rounded-md"
        src="http://image/"
      />
      <div>
        <p
          class="font-semibold my-2 line-clamp-2 text-sm font-normal"
        >
          title
        </p>
        <p
          class="text-sm opacity-70 text-xs"
        >
          channelTitle
        </p>
        <p
          class="text-sm opacity-70 text-xs"
        >
          just now
        </p>
      </div>
    </li>
    <li
      class="flex gap-2 "
    >
      <img
        alt="title2"
        class="cursor-pointer w-40 rounded-md"
        src="http://image2/"
      />
      <div>
        <p
          class="font-semibold my-2 line-clamp-2 text-sm font-normal"
        >
          title2
        </p>
        <p
          class="text-sm opacity-70 text-xs"
        >
          channelTitle2
        </p>
        <p
          class="text-sm opacity-70 text-xs"
        >
          just now
        </p>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
